### ISSUE TITLE
No method to set the timeout for distributed Gloo backend

### ISSUE THREAD

**Issue by ejoebstl** [pytorch/pytorch#6419]
Gloo has a default timeout of 30 seconds, which can be too low depending on the application. 

Currently, it's impossible to change this timeout from Pytorch, since the corresponding [DataChannel](https://github.com/pytorch/pytorch/blob/master/torch/lib/THD/base/data_channels/DataChannelGloo.cpp) does not expose any method to do so, even though Gloo has a [setTimeout](https://github.com/facebookincubator/gloo/blob/master/gloo/context.cc) method. 

I would love to help with this issue, if I can get some guidance on how to properly expose the change. Possibilities could be: 
* Providing a `set_timeout` function in `torch.distributed`
* Add an optional `timeout` argument to `torch.distributed.init_process_group`

Related to https://github.com/pytorch/pytorch/issues/3615.

**Comment by apaszke** [comment]
I think having it as an argument to `init_process_group` would be nice. To implement this, you'd need to change the [Python bindings](https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/Module.cpp#L89-L117) and [this part](https://github.com/pytorch/pytorch/blob/master/torch/lib/THD/base/DataChannel.cpp#L19-L58) to pass the argument to the `DataChannel` (think of it as a wrapper for the backend). Thanks!

**Comment by ailzhang** [comment]
Hi @ejoebstl , just checking if you are still working on this issue. Thanks!

**Comment by shudas** [comment]
Hi has there been any progress made on this issue? I'm facing read timeout issue in my distributed training job as well.

**Comment by SinghJasdeep** [comment]
^ would love to know if this has been resolved. :)

**Comment by ailzhang** [comment]
@pietern Is this available already on c10d? If not, maybe we should take it over and add it.

**Comment by ailzhang** [comment]
@pietern I will remove the "awaiting response" label and pass the task to you for now. Let me know if you need any help! :)

**Comment by pietern** [comment]
With #14435 there is a timeout kwarg you can pass to `init_process_group` to override the default timeout.
